docker/configure_workers_and_start.py

```diff
@@ -389,29 +554,219 @@ def generate_base_homeserver_config() -> None:
     subprocess.run(["/usr/local/bin/python", "/start.py", "migrate_config"], check=True)
 
 
+def parse_worker_types_from_env(
+    requested_worker_types: List[str],
+) -> Dict[str, Dict[str, Any]]:
+    """Read the desired list of worker from environment variables and prepare the data
+    for use in generating worker config files while also checking for potential gotchas.
+
+    Args:
+        requested_worker_types: The string pulled from the environment containing
+            the worker requested data.
+
+    Returns: A  dict of all information needed to generate worker files. Format:
+        {'worker_name':
+            {'worker_base_name': 'base_name'},
+
+            {'worker_types_set':
+                Set{'worker_type', 'other_worker_type'}
+            }
+        }
+    """
+    # Checking performed:
+    #   1. If a requested name contains a space
+    #   2. If a requested name contains either kind of quote mark
+    #   3. If a requested name ends with a digit
+
+    # A counter of worker_type -> int. Used for determining the name for a given
+    # worker type(s) when generating its config file, as each worker's name is just
+    # worker_(name|type(s)) + instance #
+    worker_type_counter: Dict[str, int] = defaultdict(int)
+
+    # Similar to above, but more finely grained. This is used to determine we don't have
+    # more than a single worker for cases where multiples would be bad(e.g. presence).
+    worker_type_shard_counter: Dict[str, int] = defaultdict(int)
+
+    # Dict of worker name's. This is used to check that a name requested doesn't clash
+    # with an existing name in the context of a differing worker_type, as it will error
```

Comment:
Hence the example one more line down. Preventing dumb things, like naming a worker a `worker_type` that was also requested or naming two different `worker_types` "bob". `WORKER_TYPES='bob=pusher, bob=federation_sender'` is my main thing to avoid. Should I add more to the comment?

Reply:
I think I meant: it doesn't really matter that it causes an error, the more obvious issue is that having two workers with the same name is confusing, even if that wouldn't cause an error

Reply:
Turns out it's wrong anyways. This must have been an old hold-over from an earlier version of this. I'll investigate what happened there(I suspect I used to use the keys and values in reverse), but it turns out this is now valid behavior. `WORKER_TYPES='bob=pusher, bob=federation_sender'` would create `bob1` as a pusher and `bob2` as a federation_sender. Which means you could do something really silly and name your workers `worker` for each one and it would Just Work. I will update the comments and maybe even get rid of the check altogether(after testing, of course).

Reply:
Should be fixed in b8741d743